staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/fifo_list_test.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queueset
+
+import (
+	"testing"
+)
+
+func TestFIFO(t *testing.T) {
+	tests := []struct {
+		name   string
+		testFn func(*testing.T, fifo)
+	}{
+		{
+			name: "single request in FIFO, enqueue and then dequeue",
+			testFn: func(t *testing.T, list fifo) {
+				req := &request{}
+				list.Enqueue(req)
+
+				reqGot, ok := list.Dequeue()
+				if !ok {
+					t.Errorf("expected true, but got: %t", ok)
+				}
+				if req != reqGot {
+					t.Errorf("expected dequued request: (%p), but got: (%p)", req, reqGot)
+				}
+				if list.Length() != 0 {
+					t.Errorf("expected length: %d, but got: %d)", 0, list.Length())
+				}
+			},
+		},
+		{
+			name: "multiple requests in FIFO, all requests are dequeued",
+			testFn: func(t *testing.T, list fifo) {
+				arrival := []*request{{}, {}, {}, {}, {}, {}}
+				for i := range arrival {
+					list.Enqueue(arrival[i])
+				}
+
+				dequeued := make([]*request, 0)
+				for list.Length() > 0 {
+					req, _ := list.Dequeue()
+					dequeued = append(dequeued, req)
+				}
+
+				verifyOrder(t, arrival, dequeued)
+			},
+		},
+		{
+			name: "multiple requests in FIFO, some requests are dequeued",
+			testFn: func(t *testing.T, list fifo) {
+				arrival := []*request{{}, {}, {}, {}, {}, {}}
+				half := len(arrival) / 2
+				for i := range arrival {
+					list.Enqueue(arrival[i])
+				}
+
+				dequeued := make([]*request, 0)
+				for i := 1; i <= half; i++ {
+					req, _ := list.Dequeue()
+					dequeued = append(dequeued, req)
+				}
+
+				verifyOrder(t, arrival[:half], dequeued)
+			},
+		},
+		{
+			name: "multiple requests in FIFO, remove all requests",
+			testFn: func(t *testing.T, list fifo) {
+				arrival := []*request{{}, {}, {}, {}, {}, {}}
+				removeFn := make([]removeFunc, 0)
+				for i := range arrival {
+					removeFn = append(removeFn, list.Enqueue(arrival[i]))
```

Comment:
removing the items in the right order seems like the easiest case.

I would like to verify if removing random elements works - so a test that would e.g. remove every other item instead would be better.

Reply:
I added a couple of more test:
- random order
- `remove` is idempotent